File: Exercise/app/src/main/java/com/holy/exercise/adapters/PostAdapter.java
package com.holy.exercise.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.firebase.ui.firestore.FirestoreRecyclerAdapter;
import com.firebase.ui.firestore.FirestoreRecyclerOptions;
import com.holy.exercise.R;
import com.holy.exercise.models.Post;

import java.time.LocalDate;
import java.util.Locale;


public class PostAdapter extends FirestoreRecyclerAdapter<Post, PostAdapter.ViewHolder> {

    private OnItemClickListener onItemClickListener;


    public PostAdapter(@NonNull FirestoreRecyclerOptions<Post> options) {
        super(options);
    }


    public void setOnItemClickListener(OnItemClickListener listener) {
        this.onItemClickListener = listener;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView titleText;
        private final TextView writerText;
        private final TextView dateText;
        private final ImageButton deleteButton;

        public ViewHolder(View itemView) {
            super(itemView);

            titleText = itemView.findViewById(R.id.txtPostTitle);
            writerText = itemView.findViewById(R.id.txtPostWriter);
            dateText = itemView.findViewById(R.id.txtPostDate);
            deleteButton = itemView.findViewById(R.id.ibtnDeletePost);
        }

        public void bind(Post model, OnItemClickListener listener) {

            titleText.setText(model.getTitle());
            writerText.setText(model.getWriter());

            LocalDate date = LocalDate.parse(model.getDate());
            String strDate = String.format(Locale.getDefault(),
                    "%d년 %02d월 %02d일",
                    date.getYear(), date.getMonthValue(), date.getDayOfMonth());
            dateText.setText(strDate);

            if (listener != null) {
                itemView.setOnClickListener(v -> {
                    int position = getAdapterPosition();
                    if (position != RecyclerView.NO_POSITION) {
                        listener.onItemClick(position);
                    }
                });
                deleteButton.setOnClickListener(v -> {
                    int position = getAdapterPosition();
                    if (position != RecyclerView.NO_POSITION) {
                        listener.onDeleteButtonClick(position);
                    }
                });
            }
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_post, parent, false);

        return new ViewHolder(v);
    }

    @Override
    protected void onBindViewHolder(@NonNull ViewHolder holder, int position, @NonNull Post model) {

        holder.bind(model, onItemClickListener);
    }

    public interface OnItemClickListener {
        void onItemClick(int position);
        void onDeleteButtonClick(int position);
    }

}

File: InteriorTalk/app/src/main/java/com/holy/interiortalk/adapters/PostAdapter.kt
package com.holy.interiortalk.adapters;

import android.content.Context
import android.graphics.BitmapFactory
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.firebase.ui.firestore.FirestoreRecyclerAdapter
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import com.google.firebase.storage.StorageReference
import com.holy.interiortalk.R
import com.holy.interiortalk.models.Post

class PostAdapter(
    private val context: Context,
    private val picturesRef: StorageReference,
    options: FirestoreRecyclerOptions<Post>
) : FirestoreRecyclerAdapter<Post, PostAdapter.ViewHolder>(options) {

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val postImage: ImageView = itemView.findViewById(R.id.img_post)
        private val titleText: TextView = itemView.findViewById(R.id.txt_post_title)
        private val descriptionText: TextView = itemView.findViewById(R.id.txt_post_description)
        private val writerText: TextView = itemView.findViewById(R.id.txt_post_writer)
        private val likesText: TextView = itemView.findViewById(R.id.txt_post_likes)

        fun bind(model: Post, picturesRef: StorageReference, listener: OnItemClickListener?) {

            titleText.text = model.title
            descriptionText.text = model.description
            writerText.text = model.writer

            // 이미지를 불러온다
            val megaBytes = (1024 * 1024).toLong()
            picturesRef.child("${model.id}.jpg")
                .getBytes(10 * megaBytes)
                .addOnSuccessListener {  bytes ->
                    val bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
                    postImage.setImageBitmap(bitmap)
                }
                .addOnFailureListener{
                }

            // 좋아요 수를 표시한다
            val strLikes = "+ ${model.likes}"
            likesText.text = strLikes

            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    listener?.onItemClick(model)
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {

        val itemView = LayoutInflater.from(context)
            .inflate(R.layout.item_post, parent, false)

        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, model: Post) {

        holder.bind(model, picturesRef, onItemClickListener)
    }


    interface OnItemClickListener {
        fun onItemClick(post: Post)
    }

    private var onItemClickListener: OnItemClickListener? = null

    fun setOnItemClickListener(listener: OnItemClickListener) {
        this.onItemClickListener = listener;
    }

}
